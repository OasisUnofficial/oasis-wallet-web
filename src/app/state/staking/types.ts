/* --- STATE --- */
import { ValidatorCommissionSchedule } from '../../../vendors/explorer'

export interface Validator {
  name?: string
  address: string
  escrow?: number
  status: string
  rank: number
  media?: ValidatorMediaInfo
  commission_schedule: ValidatorCommissionSchedule
  current_rate?: Rate
  current_commission_bound?: CommissionBound
}

export interface Rate {
  epochStart: number
  epochEnd?: number
  rate: number
}

export interface CommissionBound {
  lower: number
  upper: number
  epochStart: number
  epochEnd?: number
}

export interface ValidatorDetails {
  currentCommissionBounds?: CommissionBound
  scheduledCommissionBounds?: CommissionBound[]
  nextCommission?: {
    fee: number
    epochStart: number
  }
}
export interface ValidatorMediaInfo {
  website_link?: string
  email_address?: string
  twitter_acc?: string
  tg_chat?: string
  logotype?: string
}

export interface Delegation {
  validator?: Validator
  amount: string
  shares: string
  validatorAddress: string
}

export interface DebondingDelegation extends Delegation {
  epoch: number
}

export interface StakingState {
  /** List of all the validators */
  validators: Validator[]

  /** List of active delegations for the selected account */
  delegations: Delegation[]

  /** List of debonding delegations for the selected account */
  debondingDelegations: DebondingDelegation[]

  /** Addresss of the selected validator */
  selectedValidator?: string | null

  /** Lazy-loaded validator details (commission bounds and rates schedules) */
  selectedValidatorDetails: ValidatorDetails | null

  loading: boolean
}
