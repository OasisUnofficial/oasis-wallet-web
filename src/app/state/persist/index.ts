import { PayloadAction } from '@reduxjs/toolkit'
import { createSlice } from 'utils/@reduxjs/toolkit'
import {
  FinishV0MigrationPayload,
  PersistedRootState,
  PersistState,
  SetUnlockedRootStatePayload,
  StringifiedKeyWithSalt,
  UpdatePasswordPayload,
} from './types'
import { RootState } from 'types'
import { runtimeIs } from 'app/lib/runtimeIs'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { createPersistedRootReducer } from 'store/reducers'

export const STORAGE_FIELD = 'oasis_wallet_persist_v1'

/** Syncing tabs is only needed in web app, not in extension. */
export const needsSyncingTabs = runtimeIs === 'webapp'

// Simulate with `delete window.BroadcastChannel`
export const isSyncingTabsSupported = typeof BroadcastChannel === 'function'

export function getInitialState(): PersistState {
  return {
    hasPersistedProfiles: !!window.localStorage.getItem(STORAGE_FIELD),
    hasV0StorageToMigrate: false,
    // Disable persistence if tabs would override each other.
    isPersistenceUnsupported: needsSyncingTabs && !isSyncingTabsSupported,
    loading: false,
    stringifiedEncryptionKey: localStorage.getItem(STORAGE_FIELD) ? undefined : 'skipped',
    enteredWrongPassword: false,
  }
}

export const persistSlice = createSlice({
  name: 'persist',
  initialState: () => getInitialState(),
  reducers: {
    setWrongPassword(state) {
      state.enteredWrongPassword = true
      state.loading = false
    },
    resetWrongPassword(state) {
      state.enteredWrongPassword = false
    },
    setUnlockedRootState(state, action: PayloadAction<SetUnlockedRootStatePayload>) {
      /**
       * Handled in {@link createPersistedRootReducer} and {@link receivePersistedRootState}.
       * Sets `state.loading = false` and `state.stringifiedEncryptionKey`.
       */
      return
    },
    resetRootState(state) {
      /**
       * Handled in {@link createPersistedRootReducer}.
       * Sets `state.loading = false`.
       */
      return
    },
    skipUnlocking(state) {
      state.stringifiedEncryptionKey = 'skipped'
    },
    setHasV0StorageToMigrate(state, action: PayloadAction<boolean>) {
      state.hasV0StorageToMigrate = action.payload
    },

    // Handled in saga
    // ---------------
    /**
     * Set a password to encrypt state with and begin storing it to localStorage.
     */
    setPasswordAsync(state, action: PayloadAction<{ password: string }>) {
      state.loading = true
    },
    /**
     * Load encrypted state from localStorage and decrypt, or set {@link PersistState.enteredWrongPassword}.
     */
    unlockAsync(state, action: PayloadAction<{ password: string }>) {
      state.loading = true
    },
    /**
     * Remove everything from state, especially {@link PersistState.stringifiedEncryptionKey}.
     * UnlockGate should stop any components relying on missing state.
     */
    lockAsync(state) {
      state.loading = true
    },
    /**
     * Remove encrypted state from localStorage and reload.
     */
    deleteProfileAsync(state) {
      state.loading = true
    },
    updatePasswordAsync(state, action: PayloadAction<UpdatePasswordPayload>) {
      return
    },
    finishV0Migration(state, action: PayloadAction<FinishV0MigrationPayload>) {
      return
    },
  },
})

export const persistActions = persistSlice.actions

/**
 * When persisted state is unlocked use these state slices.
 */
export function receivePersistedRootState(
  prevState: RootState,
  persistedRootState: PersistedRootState,
  stringifiedEncryptionKey: StringifiedKeyWithSalt,
): RootState {
  return {
    // Explicitly list every field instead of `...prevState` to force developers
    // to consider if their new slice should be persisted.
    account: prevState.account,
    createWallet: prevState.createWallet,
    fatalError: prevState.fatalError,
    importAccounts: prevState.importAccounts,
    paraTimes: prevState.paraTimes,
    staking: prevState.staking,
    transaction: prevState.transaction,

    contacts: persistedRootState.contacts,
    evmAccounts: persistedRootState.evmAccounts,
    theme: persistedRootState.theme,
    wallet: persistedRootState.wallet,
    network: persistedRootState.network,
    persist: {
      ...getInitialState(),
      stringifiedEncryptionKey: stringifiedEncryptionKey,
    },
  }
}
