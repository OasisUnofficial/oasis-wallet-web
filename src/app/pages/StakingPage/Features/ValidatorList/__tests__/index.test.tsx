import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { stakingActions } from 'app/state/staking'
import * as React from 'react'
import { Provider } from 'react-redux'
import { configureAppStore } from 'store/configureStore'

import { ValidatorList } from '..'

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: str => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    }
  },
}))

const renderComponent = store =>
  render(
    <Provider store={store}>
      <ValidatorList />
    </Provider>,
  )

describe('<ValidatorList  />', () => {
  let store: ReturnType<typeof configureAppStore>

  beforeEach(() => {
    store = configureAppStore()
  })

  it('should match snapshot', () => {
    store.dispatch(stakingActions.updateValidators([]))
    const component = renderComponent(store)
    expect(component.container.firstChild).toMatchSnapshot()
  })

  it('should display validator details on click', async () => {
    renderComponent(store)
    store.dispatch(
      stakingActions.updateValidators([
        {
          address: 'oasis1qpc4ze5zzq3aa5mu5ttu4ku4ctp5t6x0asemymfz',
          commission_schedule: {},
          current_rate: {
            epochStart: 0,
            rate: 0.1,
          },
          rank: 0,
          status: 'active',
          escrow: 1000,
          name: 'test-validator',
          media: {
            email_address: 'test@test.com',
            tg_chat: 'telegram',
            twitter_acc: 'https://twitter.com/my_twitter',
            website_link: 'https://test.com',
          },
        },
      ]),
    )

    let row = screen.getByText(/test-validator/)
    expect(row).toBeVisible()
    userEvent.click(row)

    const details = await screen.findByTestId('validator-item')
    row = screen.getAllByText(/test-validator/)[0]

    expect(details).toBeVisible()
    userEvent.click(row)
    await waitFor(() => expect(details).not.toBeVisible())
  })

  it('should only display the details of a single validator', async () => {
    renderComponent(store)
    store.dispatch(
      stakingActions.updateValidators([
        {
          address: 'oasis1qpc4ze5zzq3aa5mu5ttu4ku4ctp5t6x0asemymfz',
          commission_schedule: {},
          current_rate: {
            epochStart: 0,
            rate: 0.1,
          },
          rank: 0,
          status: 'active',
          escrow: 1000,
          name: 'test-validator1',
          media: {
            email_address: 'test@test.com',
            tg_chat: 'telegram',
            twitter_acc: 'https://twitter.com/my_twitter',
            website_link: 'https://test.com',
          },
        },
        {
          address: 'oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe',
          commission_schedule: {},
          current_rate: {
            epochStart: 0,
            rate: 0.2,
          },
          rank: 1,
          status: 'inactive',
          escrow: 1000,
          name: 'test-validator2',
          media: {
            email_address: 'test@test.com',
            tg_chat: 'telegram',
            twitter_acc: 'https://twitter.com/my_twitter',
            website_link: 'https://test.com',
          },
        },
      ]),
    )

    let row = screen.getByText(/test-validator1/)
    expect(row).toBeVisible()
    userEvent.click(row)

    let details = await screen.findByTestId('validator-item')
    expect(screen.getByTestId('validator-item-name').textContent).toEqual('test-validator1')
    expect(details).toBeVisible()

    row = screen.getByText(/test-validator2/)
    userEvent.click(row)

    details = screen.getByTestId('validator-item')
    await waitFor(() =>
      expect(screen.getByTestId('validator-item-name').textContent).toEqual('test-validator2'),
    )
    expect(details).toBeVisible()
  })
})
