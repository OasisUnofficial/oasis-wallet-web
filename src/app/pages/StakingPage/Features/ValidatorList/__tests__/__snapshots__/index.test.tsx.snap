// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ValidatorList  /> empty should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #FFFFFF;
  color: #16213E;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 24px;
}

.c4 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  max-width: 100%;
  color: unset;
  background-color: unset;
}

.c2 {
  position: relative;
  width: 100%;
  border-radius: inherit;
  overflow-x: auto;
  overflow-y: hidden;
  min-height: 0;
}

.c3 {
  position: relative;
  width: 100%;
  display: table;
}

.c5 {
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: unset;
  background-color: unset;
}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {
  .c1 {
    padding: 12px;
  }
}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {

}

<section
  class="c0"
  data-testid="validators-list"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4 rdt_Table"
          role="table"
        >
          <div
            class="c5"
          >
            <div
              style="padding: 24px;"
            >
              There are no records to display
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`<ValidatorList  /> list should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #FFFFFF;
  color: #16213E;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 24px;
}

.c12 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 20px;
  height: 20px;
  fill: currentColor;
  stroke: currentColor;
}

.c12 g {
  fill: inherit;
  stroke: inherit;
}

.c12 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c12 *[stroke*='#'],
.c12 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c12 *[fill-rule],
.c12 *[FILL-RULE],
.c12 *[fill*='#'],
.c12 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c18 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 20px;
  height: 20px;
  fill: #2ad5ab;
  stroke: #2ad5ab;
}

.c18 g {
  fill: inherit;
  stroke: inherit;
}

.c18 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c18 *[stroke*='#'],
.c18 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c18 *[fill-rule],
.c18 *[FILL-RULE],
.c18 *[fill*='#'],
.c18 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c20 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 20px;
  height: 20px;
  fill: #d24c00;
  stroke: #d24c00;
}

.c20 g {
  fill: inherit;
  stroke: inherit;
}

.c20 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c20 *[stroke*='#'],
.c20 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c20 *[fill-rule],
.c20 *[FILL-RULE],
.c20 *[fill*='#'],
.c20 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c19 {
  font-size: 16px;
  line-height: 20px;
  color: #0500e2;
}

.c4 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  max-width: 100%;
  color: unset;
  background-color: unset;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  color: unset;
  font-size: 12px;
  font-weight: 500;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  width: 100%;
  background-color: unset;
  min-height: 52px;
  border-bottom-width: 1px;
  border-bottom-color: #AAAAAAaa;
  border-bottom-style: solid;
}

.c7 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding-left: 16px;
  padding-right: 16px;
  font-size: 16px;
}

.c17 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding-left: 16px;
  padding-right: 16px;
  word-break: break-word;
}

.c8 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 100%;
  min-width: 100px;
  min-width: 34px;
  max-width: 34px;
}

.c9 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 40ex;
  min-width: 15ex;
}

.c13 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 100%;
  min-width: 100px;
  min-width: 28ex;
  max-width: 28ex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c14 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 100%;
  min-width: 100px;
  min-width: 110px;
  max-width: 110px;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  box-sizing: border-box;
  font-size: 16px;
  font-weight: 400;
  color: unset;
  background-color: unset;
  min-height: 48px;
}

.c16:not(:last-of-type) {
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-bottom-color: #AAAAAAaa;
}

.c16:hover {
  color: unset;
  background-color: #88888833;
  -webkit-transition-duration: 0.15s;
  transition-duration: 0.15s;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  border-bottom-color: unset;
  outline-style: solid;
  outline-width: 1px;
  outline-color: unset;
}

.c16:hover {
  cursor: pointer;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: inherit;
  -webkit-justify-content: inherit;
  -ms-flex-pack: inherit;
  justify-content: inherit;
  height: 100%;
  width: 100%;
  outline: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  cursor: pointer;
}

.c10 span.__rdt_custom_sort_icon__ i,
.c10 span.__rdt_custom_sort_icon__ svg {
  -webkit-transform: 'translate3d(0, 0, 0)';
  -ms-transform: 'translate3d(0, 0, 0)';
  transform: 'translate3d(0, 0, 0)';
  opacity: 0;
  color: inherit;
  font-size: 18px;
  height: 18px;
  width: 18px;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  -webkit-transition-duration: 95ms;
  transition-duration: 95ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
}

.c10 span.__rdt_custom_sort_icon__.asc i,
.c10 span.__rdt_custom_sort_icon__.asc svg {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c10:hover,
.c10:focus {
  opacity: 0.7;
}

.c10:hover span,
.c10:focus span,
.c10:hover span.__rdt_custom_sort_icon__ *,
.c10:focus span.__rdt_custom_sort_icon__ * {
  opacity: 0.7;
}

.c11 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c15 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  position: relative;
  width: 100%;
  border-radius: inherit;
  overflow-x: auto;
  overflow-y: hidden;
  min-height: 0;
}

.c3 {
  position: relative;
  width: 100%;
  display: table;
}

@media only screen and (max-width:768px) {

}

@media only screen and (max-width:768px) {
  .c1 {
    padding: 12px;
  }
}

@media only screen and (max-width:768px) {

}

@media screen and (max-width:599px) {
  .c13 {
    display: none;
  }
}

@media screen and (max-width:599px) {
  .c14 {
    display: none;
  }
}

@media only screen and (max-width:768px) {

}

<section
  class="c0"
  data-testid="validators-list"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4 rdt_Table"
          role="table"
        >
          <div
            class="c5 rdt_TableHead"
            role="rowgroup"
          >
            <div
              class="c6 rdt_TableHeadRow"
              role="row"
            >
              <div
                class="c7 c8 rdt_TableCol"
                data-column-id="icon"
                width="34px"
              />
              <div
                class="c7 c8 rdt_TableCol"
                data-column-id="status"
                width="34px"
              />
              <div
                class="c7 c9 rdt_TableCol"
                data-column-id="name"
              >
                <div
                  class="c10 rdt_TableCol_Sortable"
                  data-column-id="name"
                  data-sort-id="name"
                  role="columnheader"
                  tabindex="0"
                >
                  <div
                    class="c11"
                    data-column-id="name"
                  >
                    validator.name
                  </div>
                  <span
                    class="asc __rdt_custom_sort_icon__"
                  >
                    <svg
                      aria-label="Down"
                      class="c12"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="m2 8.35 10.173 9.823L21.997 8"
                        fill="none"
                        stroke="#000"
                        stroke-width="2"
                      />
                    </svg>
                  </span>
                </div>
              </div>
              <div
                class="c7 c13 rdt_TableCol"
                data-column-id="escrow"
                width="28ex"
              >
                <div
                  class="c10 rdt_TableCol_Sortable"
                  data-column-id="escrow"
                  data-sort-id="escrow"
                  role="columnheader"
                  tabindex="0"
                >
                  <span
                    class="asc __rdt_custom_sort_icon__"
                  >
                    <svg
                      aria-label="Down"
                      class="c12"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="m2 8.35 10.173 9.823L21.997 8"
                        fill="none"
                        stroke="#000"
                        stroke-width="2"
                      />
                    </svg>
                  </span>
                  <div
                    class="c11"
                    data-column-id="escrow"
                  >
                    validator.escrow
                  </div>
                </div>
              </div>
              <div
                class="c7 c14 rdt_TableCol"
                data-column-id="fee"
                width="110px"
              >
                <div
                  class="c10 rdt_TableCol_Sortable"
                  data-column-id="fee"
                  data-sort-id="fee"
                  role="columnheader"
                  tabindex="0"
                >
                  <span
                    class="asc __rdt_custom_sort_icon__"
                  >
                    <svg
                      aria-label="Down"
                      class="c12"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="m2 8.35 10.173 9.823L21.997 8"
                        fill="none"
                        stroke="#000"
                        stroke-width="2"
                      />
                    </svg>
                  </span>
                  <div
                    class="c11"
                    data-column-id="fee"
                  >
                    validator.fee
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="c15 rdt_TableBody"
            role="rowgroup"
          >
            <div
              class="c16 rdt_TableRow"
              id="row-oasis1qpc4ze5zzq3aa5mu5ttu4ku4ctp5t6x0asemymfz"
              role="row"
            >
              <div
                class="c17 c8 rdt_TableCell"
                data-column-id="icon"
                data-tag="allowRowEvents"
                id="cell-icon-oasis1qpc4ze5zzq3aa5mu5ttu4ku4ctp5t6x0asemymfz"
                role="cell"
                width="34px"
              >
                <div />
              </div>
              <div
                class="c17 c8 rdt_TableCell"
                data-column-id="status"
                data-tag="allowRowEvents"
                id="cell-status-oasis1qpc4ze5zzq3aa5mu5ttu4ku4ctp5t6x0asemymfz"
                role="cell"
                width="34px"
              >
                <div>
                  <svg
                    aria-label="Status is okay"
                    class="c18"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10zM7 12l4 3 5-7"
                      fill="none"
                      stroke="#000"
                      stroke-width="2"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="c17 c9 rdt_TableCell"
                data-column-id="name"
                data-tag="allowRowEvents"
                id="cell-name-oasis1qpc4ze5zzq3aa5mu5ttu4ku4ctp5t6x0asemymfz"
                role="cell"
              >
                <div>
                  test-validator1
                </div>
              </div>
              <div
                class="c17 c13 rdt_TableCell"
                data-column-id="escrow"
                data-tag="allowRowEvents"
                id="cell-escrow-oasis1qpc4ze5zzq3aa5mu5ttu4ku4ctp5t6x0asemymfz"
                role="cell"
                width="28ex"
              >
                <div>
                  <span>
                    <div
                      class="c0"
                      style="display: inline-flex; white-space: nowrap; font-family: "Roboto mono", monospace; letter-spacing: 0;"
                    >
                      0
                    </div>
                    <span
                      class="c19"
                    >
                      <span
                        class="notranslate"
                        translate="no"
                      >
                         ROSE
                      </span>
                    </span>
                  </span>
                </div>
              </div>
              <div
                class="c17 c14 rdt_TableCell"
                data-column-id="fee"
                data-tag="allowRowEvents"
                id="cell-fee-oasis1qpc4ze5zzq3aa5mu5ttu4ku4ctp5t6x0asemymfz"
                role="cell"
                width="110px"
              >
                <div>
                  10%
                </div>
              </div>
            </div>
            <div
              class="c16 rdt_TableRow"
              id="row-oasis1qrfe9n26nq3t6vc9hlu9gnupwf4rm6wr0uglh3r7"
              role="row"
            >
              <div
                class="c17 c8 rdt_TableCell"
                data-column-id="icon"
                data-tag="allowRowEvents"
                id="cell-icon-oasis1qrfe9n26nq3t6vc9hlu9gnupwf4rm6wr0uglh3r7"
                role="cell"
                width="34px"
              >
                <div />
              </div>
              <div
                class="c17 c8 rdt_TableCell"
                data-column-id="status"
                data-tag="allowRowEvents"
                id="cell-status-oasis1qrfe9n26nq3t6vc9hlu9gnupwf4rm6wr0uglh3r7"
                role="cell"
                width="34px"
              >
                <div>
                  <svg
                    aria-label="Status is unknown"
                    class="c20"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M2 3.99C2 2.892 2.898 2 3.99 2h16.02C21.108 2 22 2.898 22 3.99v16.02c0 1.099-.898 1.99-1.99 1.99H3.99A1.995 1.995 0 0 1 2 20.01V3.99zM12 15v-1c0-1 0-1.5 1-2s2-1 2-2.5c0-1-1-2.5-3-2.5s-3 1.264-3 3m3 6v2"
                      fill="none"
                      stroke="#000"
                      stroke-width="2"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="c17 c9 rdt_TableCell"
                data-column-id="name"
                data-tag="allowRowEvents"
                id="cell-name-oasis1qrfe9n26nq3t6vc9hlu9gnupwf4rm6wr0uglh3r7"
                role="cell"
              >
                <div>
                  test-validator3
                </div>
              </div>
              <div
                class="c17 c13 rdt_TableCell"
                data-column-id="escrow"
                data-tag="allowRowEvents"
                id="cell-escrow-oasis1qrfe9n26nq3t6vc9hlu9gnupwf4rm6wr0uglh3r7"
                role="cell"
                width="28ex"
              >
                <div>
                  <span>
                    <div
                      class="c0"
                      style="display: inline-flex; white-space: nowrap; font-family: "Roboto mono", monospace; letter-spacing: 0;"
                    >
                      0
                    </div>
                    <span
                      class="c19"
                    >
                      <span
                        class="notranslate"
                        translate="no"
                      >
                         ROSE
                      </span>
                    </span>
                  </span>
                </div>
              </div>
              <div
                class="c17 c14 rdt_TableCell"
                data-column-id="fee"
                data-tag="allowRowEvents"
                id="cell-fee-oasis1qrfe9n26nq3t6vc9hlu9gnupwf4rm6wr0uglh3r7"
                role="cell"
                width="110px"
              >
                <div>
                  20%
                </div>
              </div>
            </div>
            <div
              class="c16 rdt_TableRow"
              id="row-oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe"
              role="row"
            >
              <div
                class="c17 c8 rdt_TableCell"
                data-column-id="icon"
                data-tag="allowRowEvents"
                id="cell-icon-oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe"
                role="cell"
                width="34px"
              >
                <div />
              </div>
              <div
                class="c17 c8 rdt_TableCell"
                data-column-id="status"
                data-tag="allowRowEvents"
                id="cell-status-oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe"
                role="cell"
                width="34px"
              >
                <div>
                  <svg
                    aria-label="Status is critical"
                    class="c20"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12.703 2.703a.99.99 0 0 0-1.406 0l-8.594 8.594a.99.99 0 0 0 0 1.406l8.594 8.594a.99.99 0 0 0 1.406 0l8.594-8.594a.99.99 0 0 0 0-1.406l-8.594-8.594zM8.983 14.7 14.7 8.983m-5.717 0L14.7 14.7"
                      fill="none"
                      stroke="#000"
                      stroke-width="2"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="c17 c9 rdt_TableCell"
                data-column-id="name"
                data-tag="allowRowEvents"
                id="cell-name-oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe"
                role="cell"
              >
                <div>
                  test-validator2
                </div>
              </div>
              <div
                class="c17 c13 rdt_TableCell"
                data-column-id="escrow"
                data-tag="allowRowEvents"
                id="cell-escrow-oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe"
                role="cell"
                width="28ex"
              >
                <div>
                  <span>
                    <div
                      class="c0"
                      style="display: inline-flex; white-space: nowrap; font-family: "Roboto mono", monospace; letter-spacing: 0;"
                    >
                      0
                    </div>
                    <span
                      class="c19"
                    >
                      <span
                        class="notranslate"
                        translate="no"
                      >
                         ROSE
                      </span>
                    </span>
                  </span>
                </div>
              </div>
              <div
                class="c17 c14 rdt_TableCell"
                data-column-id="fee"
                data-tag="allowRowEvents"
                id="cell-fee-oasis1qzyqaxestzlum26e2vdgvkerm6d9qgdp7gh2pxqe"
                role="cell"
                width="110px"
              >
                <div>
                  20%
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
